v2/health.go

```diff
@@ -0,0 +1,127 @@
+	subscription string
+	// interval is the time between health checks.
+	interval time.Duration
+	options  *HealthCheckerOptions
+}
+
+// HealthCheckerOptions configures the HealthChecker.
+// HealthCheckTimeout defaults to HealthChecker.interval if not set or set to 0 or set to be larger than interval.
+type HealthCheckerOptions struct {
+	// HealthCheckTimeout is the context timeout for each health check
+	HealthCheckTimeout time.Duration
+}
+
+// NewHealthChecker creates a new HealthChecker with the provided clients, entity, subscription, and interval.
+// clients is a map of namespace name to azservicebus.Client.
+func NewHealthChecker(clients map[string]*azservicebus.Client, entity, subscription string, interval time.Duration, options *HealthCheckerOptions) *HealthChecker {
+	if options == nil {
+		options = &HealthCheckerOptions{}
+	}
+	if options.HealthCheckTimeout == 0 || options.HealthCheckTimeout > interval {
+		options.HealthCheckTimeout = interval
+	}
+
+	return &HealthChecker{
+		clients:      clients,
+		entity:       entity,
+		subscription: subscription,
+		interval:     interval,
+		options:      options,
+	}
+}
+
+// StartSenderPeriodicHealthCheck starts a periodic health check for the sender.
+// It uses azservicebus.Sender.NewMessageBatch() to perform the health check.
+// Stops when the context is cancelled.
+func (h *HealthChecker) StartSenderPeriodicHealthCheck(ctx context.Context) {
+	for namespace, client := range h.clients {
+		go h.periodicHealthCheck(ctx, h.senderHealthCheck, namespace, client)
+	}
+}
+
+// StartReceiverPeriodicHealthCheck starts a periodic health check for the receiver
+// It uses azservicebus.Receiver.PeekMessages() to perform the health check.
+// Stops when the context is cancelled.
+func (h *HealthChecker) StartReceiverPeriodicHealthCheck(ctx context.Context) {
+	for namespace, client := range h.clients {
+		go func(namespace string, client *azservicebus.Client) {
+			go h.periodicHealthCheck(ctx, h.receiverHealthCheck, namespace, client)
+		}(namespace, client)
+	}
+}
+
+func (h *HealthChecker) periodicHealthCheck(ctx context.Context, healthCheckFunc HealthCheckFunc, namespace string, client *azservicebus.Client) {
+	nextCheck := time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Until(nextCheck)):
+			if err := healthCheckFunc(ctx, namespace, client); err != nil {
+				log(ctx, err)
+			}
+			nextCheck = nextCheck.Add(h.interval)
+		}
+	}
+}
+
+func (h *HealthChecker) senderHealthCheck(ctx context.Context, namespace string, client *azservicebus.Client) error {
+	s, err := client.NewSender(h.entity, nil)
+	if err != nil {
+		sender.Metric.IncHealthCheckFailureCount(namespace, h.entity)
+		return err
+	}
+	sbCtx, cancel := context.WithTimeout(ctx, h.options.HealthCheckTimeout)
+	defer cancel()
+	_, err = s.NewMessageBatch(sbCtx, nil)
+	if err != nil {
+		sender.Metric.IncHealthCheckFailureCount(namespace, h.entity)
```

Comment:
can we move the metric handling iton the calling func? or is that then making the select too messy?